Show UPE payment methods in saved order on block checkout page

Mayisha opened the pull request:
Fixes #2994 

In https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3184 we have implemented the ability to order the payment methods on the Stripe settings page when UPE is enabled but the order does not reflect on the block checkout. In this PR, I have fixed that and the order of the UPE methods on block checkout is now the same as Stripe settings page.

## Changes proposed in this Pull Request:
- In [blocks code](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Blocks/Payments/Api.php#L86), the payment methods are retrieved from `WC()->payment_gateways->payment_gateways()` (all gateways). For that, the `woocommerce_available_payment_gateways` filter does not work on the block checkout for reordering the payment methods.
- I am now updating the `woocommerce_gateway_order` option which holds all payment gateway order when the Stripe payment methods order is changed. In `add_stripe_methods_in_woocommerce_gateway_order` we find the main Stripe gateway position in `woocommerce_gateway_order` and push all the other Stripe methods there in the order saved in the Stripe settings page. 
- `WC()->payment_gateways->payment_gateways()` returns the list based on the order of `woocommerce_gateway_order`, so now the order of payment methods on blocks checkout reflects the order in the Stripe settings page.

## Testing instructions
- Enable the new checkout experience.
- On the Stripe settings page, confirm that there is a `Change display order` button at the top of the payment method list.
- Change the order of a few methods and save changes.
- As a shopper, add product to your cart and go to the block checkout page. Confirm that the Stripe methods are displayed in the saved order.
- Go to the shortcode checkout page and confirm that the order is correct here as well.
- In your admin dashboard, go to **WooCommerce > Settings > Payments**. Change the order of a few gateways and enable one or two gateways along with Stripe.
- Check the shortcode and block checkout pages and confirm that the payment method order is correct.

Mayisha committed:
show sortable list for UPE

Mayisha committed:
save and show ordered payment methods for UPE

Mayisha committed:
Merge branch 'develop' into task/2994-change-display-order-upe

Mayisha committed:
refactor the helper method

Mayisha committed:
saved enabled payment methods in order

Mayisha committed:
show the upe methods on the checkout page in the same order of the settings page

Mayisha committed:
Merge branch 'develop' into task/2994-change-display-order-upe

Mayisha committed:
add changelog

Mayisha committed:
remove unnecessary code

Mayisha committed:
fix tests

Mayisha committed:
use hardcoded enum

Mayisha committed:
copy legacy order to upe order when first swtched to the new checkout experience

Mayisha committed:
add check

Mayisha committed:
Merge branch 'develop' into task/2994-change-display-order-upe

Mayisha committed:
Merge branch 'develop' into task/2994-change-display-order-upe

Mayisha committed:
dynamically get gateways as before in enum

Mayisha committed:
fix error
- use correct method name after merging in develop
- copy over the legacy order without stripe_ prefix

Mayisha committed:
save the enabled payment methods in the new order on reorder in settings

Mayisha committed:
update the order in 'woocommerce_gateway_order' option when a new order is saved from the stripe settings page

Mayisha committed:
move method to helper class

Mayisha committed:
set stripe gateways in the gateway order list when gateways are reordered on the payment settings page

Mayisha committed:
Merge branch 'develop' into task/2994-upe-display-order-blocks

a-danae commented on includes/class-wc-stripe-helper.php:
```diff
@@ -722,6 +722,52 @@ public static function get_upe_settings_enabled_payment_method_ids( $gateway ) {
+				$ordered_available_stripe_methods[] = $gateway_id;
+				continue;
+			} elseif ( 'sepa_debit' === $payment_method_id ) {
+				$gateway_id = 'stripe_sepa';
```
Is this `$gateway_id` used? Looks like this definition of `$gateway_id` is overwritten in 745.

Mayisha replied:
Right. Moved that line above in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3222/commits/741a24c6fcc446b087f6cf6e9bb593278549520e

a-danae commented on includes/class-wc-stripe-helper.php:
```diff
@@ -722,6 +722,52 @@ public static function get_upe_settings_enabled_payment_method_ids( $gateway ) {
+				$gateway_id = 'stripe_sepa';
+			}
+
+			$gateway_id = str_starts_with( $payment_method_id, 'stripe_' ) ? $payment_method_id : 'stripe_' . $payment_method_id;
```
When Legacy is enabled , we're adding a `stripe_stripe`. Is this something we should adjust?

<img width="400" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/ad8e36a1-d0d2-4c90-8362-f6089598a5d1">

Mayisha replied:
Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3222/commits/741a24c6fcc446b087f6cf6e9bb593278549520e

a-danae commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -311,6 +321,8 @@ function ( $id ) {
 			$this->gateway->update_option( 'stripe_legacy_method_order', $ordered_payment_method_ids );
 		}
 
+		WC_Stripe_Helper::add_stripe_methods_in_woocommerce_gateway_order( $ordered_payment_method_ids );
```
We're updating the payment methods order on Legacy using the `woocommerce_available_payment_gateways` filter, right? Curious what's the use of `WC_Stripe_Helper::add_stripe_methods_in_woocommerce_gateway_order` in that case.

Mayisha replied:
The reordering in `woocommerce_available_payment_gateways` is not working on the payment methods order on the blocks checkout page. 
In shortcode checkout page, the payment methods are [displayed](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-template-functions.php#L2410) from `get_available_payment_gateways` which uses [this filter](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-payment-gateways.php#L333).
In [blocks code](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Blocks/Payments/Api.php#L86) the payment method order is taken from `WC()->payment_gateways->payment_gateways()` (all gateways). not from `get_available_payment_gateways`. As the `WC()->payment_gateways->payment_gateways()` returns the list in order from `woocommerce_gateway_order` option, I am updating this option every time we save the order in Stripe settings page or the main payments settings page with the help of `WC_Stripe_Helper::add_stripe_methods_in_woocommerce_gateway_order`.

a-danae replied:
Makes sense. Thanks for the detailed description @Mayisha ! 🙂

a-danae commented on includes/class-wc-stripe-helper.php:
```diff
@@ -722,6 +722,52 @@ public static function get_upe_settings_enabled_payment_method_ids( $gateway ) {
+				$gateway_id = 'stripe_sepa';
+			}
+
+			$gateway_id = str_starts_with( $payment_method_id, 'stripe_' ) ? $payment_method_id : 'stripe_' . $payment_method_id;
```
Is this check for whether `$payment_method_id` starts with `stripe_` used for legacy? Asking because I don't see any `$payment_method_id` to start with `stripe_`, but they do on legacy.

If so, would it be better to use the array passed to this method before we [prepend](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3222/files#diff-679186f0c3e522393b5394f6e65890f1b2bff5ef64efb250eee90d6348dc841eR316) `stripe_` to the array values?

Mayisha replied:
> Is this check for whether $payment_method_id starts with stripe_ used for legacy? Asking because I don't see any $payment_method_id to start with stripe_, but they do on legacy.

This check is for new experience (UPE). In legacy the method names already have `stripe_` but not in UPE. The `woocommerce_gateway_order` option is supposed to keep the gateway names with `stripe_`.

> If so, would it be better to use the array passed to this method before we [prepend](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3222/files#diff-679186f0c3e522393b5394f6e65890f1b2bff5ef64efb250eee90d6348dc841eR316) stripe_ to the array values?

We are also calling this function [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3222/files#diff-4c2be2e009983410faf0c92dfb7770331510dfc470e5c156d5055fbf46f324f5R56) from `set_stripe_gateways_in_list` function. If we do the mapping before passing to the function, we will have to do it twice.

a-danae reviewed (changes requested):
Thanks for addressing this issue @Mayisha ! 

✅ The payment methods are ordered as expected on the block checkout on this branch.
✅ Multibanco isn't being displayed on this branch, but it works as expected when merging `develop`.

❓ When reordering the payment methods from the WooCommerce page, Stripe is always at the bottom of the list. Would we please double-check this? It's working fine in `develop`.

https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/1a80935c-b3da-4d63-8d3e-190d59075d22

I left a few comments in the code, but those are just questions. Requesting changes because of the behavior above.

Mayisha committed:
remove unnecessary code

Mayisha committed:
fix gateway name maping in 'add_stripe_methods_in_woocommerce_gateway_order' function

Mayisha commented:
Hi @a-danae 👋 

Thanks for the feedback I have addressed your comments and made a couple of changes in the last 2 commits. 

> When reordering the payment methods from the WooCommerce page, Stripe is always at the bottom of the list. Would we please double-check this? It's working fine in develop.

I can not reproduce this issue. Could you try again and test this? May be the latest changes have fixed this as well.

a-danae commented on includes/admin/class-wc-stripe-settings-controller.php:
```diff
@@ -40,6 +40,22 @@ public function __construct( WC_Stripe_Account $account, WC_Stripe_Payment_Gatew
+		add_action( 'update_option_woocommerce_gateway_order', [ $this, 'set_stripe_gateways_in_list' ] );
+	}
+
+	public function set_stripe_gateways_in_list( $ordering ) {
```
It'd probably be good to have some documentation for this function. What do you think?

Mayisha replied:
Added in [bec337b](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3222/commits/bec337bcd4c032523d93d5545bc1de15e86a48ed)

a-danae reviewed (commented):
Thanks for the updates here, @Mayisha ! I marked all the previous conversations as resolved.

> I can not reproduce this issue. Could you try again and test this? May be the latest changes have fixed this as well.

I'm not able to reproduce it over here again, not even in older commits... Strange but we can dive there if it happens again and I get it consistently on another test site.

❓ Looks like the order isn't correct on the block checkout. Notice that SEPA should be second, but it's after Check Payments instead. Would you please confirm if it happens on your end?

| Block checkout | Shortcode checkout | Settings page |
|--------|--------|--------|
| <img width="600" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/6b99c0f6-cdb9-45f1-9cbb-9284473ce8f5"> |<img width="600" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/ad18439a-b7ca-49ee-9cda-fa8a498e6c37">|<img width="600" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/390a3723-cbed-427f-9b8d-c1252e3a3772">|

The order of the payment methods doesn't seem to be wrong when [updating it](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3222/files#diff-dbc2cfbfe4b247c27c20d86922819977adc6f7723c0b047cb9722a2f47b7ee73R764).

<img width="300" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/965c9f83-b709-4d87-bc30-4f5f3ea228ca">

But for some reason, this order isn't preserved when retrieving the payment gateways to be displayed on [the checkout block](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Blocks/Payments/Api.php#L88)
<img width="300" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/9ac62114-dd23-44c5-9ec2-bf28117c51be">

The key of the SEPA gateway is 1001 on my end in [this line](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-payment-gateways.php#L123), with the payment methods I have enabled.

<img width="300" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/3a6d5a2f-1e64-4a8a-84c2-c4672e7f3e53">


I recall this was working as expected before, so maybe it was a regression from the latest commits? I also tried re-reordering the payment methods without luck.

a-danae commented on includes/class-wc-stripe-helper.php:
```diff
@@ -722,6 +722,48 @@ public static function get_upe_settings_enabled_payment_method_ids( $gateway ) {
+		$ordered_available_stripe_methods = [];
+		// Map the Stripe payment method list to the right format to save in the 'woocommerce_gateway_order' option.
+		foreach ( $ordered_payment_method_ids as $payment_method_id ) {
+			$gateway_id = str_starts_with( $payment_method_id, 'stripe' ) ? $payment_method_id : 'stripe_' . $payment_method_id;
```
Just noticed this (I did the same on a PR of mine 😅 ). We should use an alternative for the `str_starts_with` function because it was introduced in [PHP 8.0](https://www.php.net/manual/en/function.str-starts-with.php#:~:text=str_starts_with-,(PHP%208),-str_starts_with%20%E2%80%94%20Checks), and we should support previous PHP versions. [strpos](https://www.php.net/manual/en/function.strpos.php) could do.

Mayisha replied:
Thanks for raising this 🤦‍♀️ 
Fixed this in [bec337b](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3222/commits/bec337bcd4c032523d93d5545bc1de15e86a48ed)

Mayisha committed:
Merge branch 'develop' into task/2994-upe-display-order-blocks

Mayisha committed:
Merge branch 'develop' into task/2994-upe-display-order-blocks

Mayisha committed:
address feedback

Mayisha committed:
fix sepa showing up in wrong order

Mayisha committed:
remove unnesessary function

Mayisha committed:
add stripe methods to 'woocommerce_gateway_order' option on plugin update

Mayisha committed:
add changelog

Mayisha committed:
Merge branch 'develop' into task/2994-upe-display-order-blocks

Mayisha committed:
maintain correct order on upe toggle

Mayisha committed:
fix undefined index

Mayisha commented:
Hi @a-danae 👋 

Thanks for the detailed feedback. I have addressed the two pieces of feedback about adding documentation of the function and using `strpos` in [bec337b](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3222/commits/bec337bcd4c032523d93d5545bc1de15e86a48ed).

> Looks like the order isn't correct on the block checkout. Notice that SEPA should be second, but it's after Check Payments instead. Would you please confirm if it happens on your end?

I could reproduce this and this was an issue specific to SEPA only. Fixed in [8c19fbb](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3222/commits/8c19fbb24e6bdc11f58e010a039d121763517b99).

> The key of the SEPA gateway is 1001 on my end in [this line](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-payment-gateways.php#L123), with the payment methods I have enabled.

I can not reproduce this one. 😩 

I have made some other changes in my last few commits.
- I am updating the `woocommerce_gateway_order` option ( which holds all the gateway order from _WooCommerce > Settings > Payments_ page) and inserting the Stripe methods in this option in the correct position. This is done in the helper method `add_stripe_methods_in_woocommerce_gateway_order`. The reason for adding the Stripe methods in the `woocommerce_gateway_order` is explained in the description.
- Removed the usage of `woocommerce_available_payment_gateways` filter and the corresponding callback function `reorder_available_payment_gateways`. This can be removed as I am now adding the Stripe methods in the `woocommerce_gateway_order` always.
- Calling the `add_stripe_methods_in_woocommerce_gateway_order` function to insert the Stripe methods to the correct position (the position of the main Stripe gateway on the _WooCommerce > Settings > Payments_ page) in the following scenarios.
    - When we reorder the payment methods in the Stripe settings page.
    - When a new gateway order is saved from _WooCommerce > Settings > Payments_ page. 
    - When UPE flag is toggled as UPE and non-UPE settings might have different order in the Stripe settings page.
    - When plugin is updated. In the previous version, we did not include the Stripe methods in the 
      `woocommerce_gateway_order` option. 

Apologies for the back and forth here 😬

a-danae committed:
Merge branch 'develop' into task/2994-upe-display-order-blocks

a-danae commented on includes/class-wc-stripe-helper.php:
```diff
@@ -717,6 +717,61 @@ public static function get_upe_settings_enabled_payment_method_ids( $gateway ) {
+				continue; // Skip the other stripe gateways. We'll add all Stripe methods back in the right order.
+			} elseif ( 'stripe' === $gateway ) {
+				unset( $gateway_order['stripe'] );
+				foreach ( $ordered_available_stripe_methods as $ordered_available_stripe_method ) {
```
What do you think about adding comments explaining why this nested loop is needed? It may be just me but I had to go through it a few times to understand what was happening.

Mayisha replied:
Added in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3222/commits/b191fbcd6201462a74a9e81a22094f3382125366 🎉

a-danae reviewed (approved):
Thanks for iterating on the approach here @Mayisha! Things look good to me 🎉 

✅ SEPA is now ordered as expected. I can't replicate the behavior I described anymore.
✅ Payment methods are ordered as expected in both the shortcode and block checkouts.
✅ The payment methods on Legacy continue to behave as expected.
✅ The displayed payment methods are consistent with the ones displayed on `develop`.
✅ The code looks good to me. I find it so much simpler after the last refactor!

Thanks again for the work and effort on this one. Happy to see this merged! 🙂

Mayisha committed:
add more comments for clarification
